### PR TITLE
Catch unhandled promise rejections & forward them to the ErrorApi

### LINKED ISSUE #4782: core: Add ErrorApi helper that catches unhandled promise rejections and forwards them
## Feature Suggestion

Add an `ErrorApi` helper that catches unhandled promise rejections and forwards them to an `ErrorApi` implementation.

## Possible Implementation

Something like an `UncaughtErrorForwarder` class that uses the [unhandledrejection  event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event#preventing_default_handling) to capture uncaught promise rejections.

Could be something like this:

```tsx
class UnhandledErrorForwarder {
  static forward(errorApi: ErrorApi) {
    ...
  }
}
```

And then we hook it up as a part of the default `ErrorApi` implementation [here](https://github.com/backstage/backstage/blob/28ac6aefb0901fe07eb580a7ed2ef7c23ef5b787/packages/core/src/api-wrappers/defaultApis.ts#L68) like so:

```ts
createApiFactory({
  api: errorApiRef,
  deps: { alertApi: alertApiRef },
  factory: ({ alertApi }) => {
    const errorApi = new ErrorAlerter(alertApi, new ErrorApiForwarder());
    UnhandledErrorForwarder.forward(errorApi);
    return errorApi;
  }
}),
```

## Context

Suggestion by @grantila that we handle unhandled rejections somehow.

### PR DESCRIPTION
**Pull request by tudi2d**
Forwarding unhandled errors to the ErrorApi using post. They are still displayed to the developer in the console & hidden from the normal user. I decided against the implementation of @grantila as he wrote in his docs, that it's not meant for production & to keep it minimal - can change this as well. Happy about review & feedback! :)

**Questions:**
- Should the error stack be logged & should the Message be hidden?
- What happens with hidden errors reported to the ErrorApi? Is there a logger observing them?

Closes #4782 

https://user-images.githubusercontent.com/8904624/121685454-3cf2aa80-cac0-11eb-981d-f3fd041a50f5.mov

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/fair-knives-relax.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/core': patch
+'@backstage/core-api': patch
+---
+
+Catch unhandled promise rejections & forward them to the ErrorApi to align with general error handling
```
Comment (Rugvip) [inline .changeset/fair-knives-relax.md:6]: Probably mention how we're introducing a new `UnhandledErrorForwarder` export and how it's installed by default. And if users wish to omit that functionality they can override the API in their app.

**Comment by Rugvip** [comment]
@tudi2d Regarding what happens to the errors, it can be customized, but by default so show them in an error alert popup at the top of the screen. Although only if the error context `hidden` is not set to true

**Comment by tudi2d** [comment]
Thanks for the detailed review - I will improve the PR! 🙂

**Comment by tudi2d** [comment]
> @tudi2d Regarding what happens to the errors, it can be customized, but by default so show them in an error alert popup at the top of the screen. Although only if the error context `hidden` is not set to true

I edited the PR (will do the commit sign-off). A question I have is, how the ErrorForwarder can be customised with the `hidden` field. I added the optional `hidden` parameter, but do I have to additional change smth. in the ErrorApi?

**Comment by benjdlambert** [comment]
Think we want to make `hidden: false` by default right? And let's users override it with some docs on how to do that?

**Comment by tudi2d** [comment]
> Think we want to make `hidden: false` by default right? And let's users override it with some docs on how to do that?

Yes. I tried to do so by setting the `hidden` parameter to `false` in `UnhandledErrorForwarder.ts`. I'm not sure how a user can override it the API though - would be happy to document it, if you can give me an example 🙂

**Comment by benjdlambert** [comment]
Maybe confirm that this works and flesh out the documentation a bit, but what about this:

In `packages/app/src/api.ts` you can add

```ts
  createApiFactory({
    api: errorApiRef,
    deps: { alertApi: alertApiRef },
    factory: ({ alertApi }) =>
      new ErrorAlerter(alertApi, new ErrorApiForwarder()),
    factory: ({ alertApi }) => {
      const errorApi = new ErrorAlerter(alertApi, new ErrorApiForwarder());
      UnhandledErrorForwarder.forward(errorApi, { hidden: false });
      return errorApi;
    },
  }),
  ```

On second thoughts too, maybe we should wrap up the `hidden` into an config object that gets passed rather than just a boolean value. Don't want to lock it to a boolean in order to extend the API over time

File: packages/core-app-api/api-report.md [hunk 2]
```diff
@@ -394,6 +394,12 @@ export type SignInResult = {
     signOut?: () => Promise<void>;
 };
 
+// @public (undocumented)
```
Comment (benjdlambert) [inline packages/core-app-api/api-report.md:397]: Think you'll need the generate these reports again as they're outdated now

File: .changeset/fair-knives-relax.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/core': patch
```
Comment (benjdlambert) [inline .changeset/fair-knives-relax.md:2]: one more thing on this that I've forgotten.

I think this wants to be `'core-app-api': patch` only as we only change that package now 👍

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**